File: Java Based Archives/TMDBJava/app/src/main/java/com/android/tmdb_java/service/RetrofitInstance.java
package com.android.tmdb_java.service;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitInstance {

    private static Retrofit retrofit = null;
    private static final String BASE_URL = "https://api.themoviedb.org/3/";

    // Method to get Retrofit service instance
    public static MovieDataService getService() {
        // If Retrofit instance is null, create new instance
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        // Return MovieDataService interface implementation
        return retrofit.create(MovieDataService.class);
    }
}

File: RetrofitDemo/app/src/main/java/com/anushka/retrofitdemo/RetrofitInstance.kt
package com.anushka.retrofitdemo

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import com.google.gson.GsonBuilder
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import java.util.concurrent.TimeUnit

/**
 * Singleton class responsible for providing a Retrofit instance.
 */
class RetrofitInstance {


    companion object {
        private const val BASE_URL = "https://jsonplaceholder.typicode.com/"
        private val interceptor = HttpLoggingInterceptor().apply {
            this.level = HttpLoggingInterceptor.Level.BODY
        }
        val client = OkHttpClient.Builder().apply{
            this.addInterceptor(interceptor)
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(20,TimeUnit.SECONDS)
                .writeTimeout(25,TimeUnit.SECONDS)
        }.build()

        /**
         * Returns a Retrofit instance configured with the base URL and Gson converter factory.
         * @return Configured Retrofit instance.
         */
        fun getRetrofitInstance(): Retrofit {
            return Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create(GsonBuilder().create()))
                .build()
        }
    }
}
